tests/group_functions/group_broadcast.h

```diff
@@ -131,87 +131,127 @@ void broadcast_sub_group(sycl::queue& queue) {
         T local_var(splat_init<T>(0));
 
+        // To simplify the test, we are only checking the first sub-group
         if (sub_group.get_group_id()[0] == 0) {
-          // find local id of last group item
-          sycl::id<1> last_item = sub_group.get_local_range();
-          --last_item[0];
-
-          // broadcast from the first workitem
-          local_var = splat_init<T>(item.get_global_linear_id() + 1);
+          // Find local id of first, last and some third sub-group item in
+          // between. Will be used to check different combinations of
+          // broadcasting and receiving work-items
+          sycl::id<1> first_id = 0;
+          sycl::id<1> mid_id = sub_group.get_local_range() / 2;
+          sycl::id<1> last_id = sub_group.get_local_range();
+          --last_id[0];
+
+          // Broadcast from the first work-item
           ASSERT_RETURN_TYPE(
-              T, sycl::group_broadcast(sub_group, local_var),
+              T, sycl::group_broadcast(sub_group, value_to_broadcast),
               "Return type of group_broadcast(sub_group g, T x) is wrong\n");
 
-          local_var = sycl::group_broadcast(sub_group, local_var);
-          if (sub_group.get_local_linear_id() ==
-              sub_group.get_local_linear_range() - 1)
-            res_acc[0] = local_var;
+          if (sub_group.leader()) {
+            // Work-item which does the broadcast stores value to broadcast to
+            // use it later as a reference
+            origin_values_acc[0] = value_to_broadcast;
+          }
+          auto broadcasted_value =
+              sycl::group_broadcast(sub_group, value_to_broadcast);
+          // We read broadcasted value in another work-item
+          if (sub_group.get_local_id() == last_id)
+            broadcasted_values_acc[0] = broadcasted_value;
 
-          // broadcast from the last workitem 1
-          local_var = splat_init<T>(item.get_global_linear_id() + 1);
+          // Broadcast from the last work-item
           ASSERT_RETURN_TYPE(
-              T, sycl::group_broadcast(sub_group, local_var, last_item),
+              T, sycl::group_broadcast(sub_group, value_to_broadcast, last_id),
               "Return type of group_broadcast(sub_group g, T x, "
               "sub_group::linear_id_type local_linear_id) is wrong\n");
 
-          local_var = sycl::group_broadcast(
-              sub_group, local_var, sub_group.get_local_linear_range() - 1);
-          if (sub_group.get_local_linear_id() == 0) res_acc[1] = local_var;
+          if (sub_group.get_local_id() == last_id) {
+            // Work-item which does the broadcast stores value to broadcast to
+            // use it later as a reference
+            origin_values_acc[1] = value_to_broadcast;
+          }
+
+          broadcasted_value =
+              sycl::group_broadcast(sub_group, value_to_broadcast,
+                                    sub_group.get_local_linear_range() - 1);
```

Comment:
Is it not possible to use `last_id` here?

Reply:
I should probably leave a comment here, because `group_broadcast` has overloads accepting both `size_t` and `sycl::it` as an index of a work-item which should be used as source.

Were are trying to test both interfaces here. I should also fix the last argument in `ASSERT_RETURN_TYPE` to match that particular overload with plain `size_t`

Reply:
The more comments I read, the happier my brain is. ;-)

Reply:
Applied all the mentioned changes in 2952e4e0fee35fe8d3d4ff9a23b5e48abd4ff311